src/main/java/com/puppycrawl/tools/checkstyle/JavadocPropertiesGenerator.java

```diff
@@ -57,7 +57,8 @@
      * The end of the sentence is determined by the symbol "period", "exclamation mark" or
      * "question mark", followed by a space or the end of the text.
      */
-    private static final Pattern END_OF_SENTENCE_PATTERN = Pattern.compile("(.*?[.?!])(\\s|$)");
+    private static final Pattern END_OF_SENTENCE_PATTERN = Pattern.compile(
+        "(([^.?!]|[.?!](?!\\s|$))*+[.?!])(\\s|$)");
```

Comment:
This new expression makes no sense to me.

> The end of the sentence is determined by the symbol "period", "exclamation mark" or "question mark", followed by a space or the end of the text.

Since this is using `find` that means it doesn't have to be an exact match.

That means the `.*?` can go away completely.

The rest has to be followed by either spaces or the end. Why can't `(\\s|$)` be replaced with `\\s*$` ?

We can't make the whole expression could be `[.?!]\\s*$` ?

Reply:
> That means the `.*?` can go away completely. `find` is already doing a search in the text.

For determining whether there's a match or not, I agree. But we also call `matcher.group(1)`. And including `.*?` in the regex changes what is captured by this first capture group.
This is because of how regexes are applied: they try to match the input String from the beginning, and only if that is not possible, they start again and try to match from the next position in the input String. This [short article](https://community.appway.com/screen/kb/article/checking-strings-avoiding-catastrophic-backtracking-1482810891360#:%7E:text=Catastrophic%20backtracking%20is%20a%20condition,the%20string%20to%20not%20match) might explain it better, and the [regex debugger](https://regex101.com/r/RFqxII/1/debugger) mentioned there visualizes what is happening.

> The rest has to be followed by either spaces or the end. Why can't `(\\s|$)` be replaced with `\\s*$` ?

The regex is used to extract the first sentence from a (potentially) longer text. This wouldn't work anymore with `\\s*$`. E.g. for input "This is a sentence. This is a second sentence.", `(([^.?!]|[.?!](?!\s|$))*+[.?!])(\s|$)` matches "This is a sentence. ", but `(([^.?!]|[.?!](?!\s|$))*+[.?!])\s*$` matches " This is a second sentence."

> We can't make the whole expression be [.?!]\\s*$ or [.?!]\\s*?$?

As discussed above, the regex is not only used to determine whether there's a match or not, but we actually are interested in the capture group, i.e. *what exactly* is matched by the regex.

> This new expression makes no sense to me.

Maybe the most important thing to know is that `(?!\s|$)` is a "negative lookahead" (see [Pattern](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html)).

`[.?!](?!\s|$)` matches a punctuation character that is followed by neither whitespace nor by the end of the input string.

`([^.?!]|[.?!](?!\s|$))*+` thus matches any characters that do not constitute the end of a sentence.

And the whole regex `(([^.?!]|[.?!](?!\s|$))*+[.?!])(\s|$)` matches (any characters that do not constitute the end of a sentence) plus (the end of a sentence). I.e. it matches exactly one sentence.